### ISSUE TITLE
Add Selective Mechanism for Feature Extraction

### ISSUE THREAD

**Issue by mikel-brostrom** [mikel-brostrom/boxmot#1707]
To all trackers

https://github.com/mikel-brostrom/boxmot/issues/1620

**Comment by Fleyderer** [comment]
I don't know where it is better to discuss, but if we follow the principle of Occam's razor, I will discuss it here 🙂

##  Current setup
I use yolox_m.pth model, pretrained by ByteTrack authors. After small fixes like changing Kalman filter from XYAH to XYWH metrics on my tracker `puretrack` (it was still 99% bytetrack) on MOT17/train are: 

HOTA: 71.8, MOTA: 87.16, IDF1: 80.3

Command:
```
python tracking/val.py --yolo-model yolox_m.pth --tracking-method puretrack --reid-model osnet_x1_0_msmt17.pt --source tracking/val_utils/data/MOT17/train --conf 0.2
```

And then I've rewrote bytetrack from scratch with maximum vectorization and decreasing unnecessary memory reallocation, adding ReID, so now it works faster than bytetrack even with cos_dist calculation and embeddings processing - up to 30-40% (of course with precalculated dets and embs).

So, now i have:

HOTA: 72.0, MOTA: 87.1, IDF1: 80.5


## Main thoughts

But after implementing selective feature extraction, or using any else new "rocket science" techniques, metrics are dropping to HOTA at least 68, sometimes they are much worse. All the same with these trackers like strongsort, which works on 5-10 FPS (and this is despite the inference time...), while bytetrack is 10x faster and results in better metrices. It makes me think that something is really wrong here - maybe with me 🙂

Of course, it could be like "ByteTrack trained these models, so they have an advantage" or anything else. But I have my own, manually annotated video from street cam only for testing purposes with no any train or finetuning, and there is the same thing - ByteTrack not just faster, but at the same time it is better, much better.

Current method with selective extraction of embeddings is quite good, but as I can see, not for two-stage trackers like bytetrack, because we calculate not only "risky" embeddings, but also for new detects and matches from newly created tracks, and we still don't know, which embeddings we have to calculate and which are unnecessary. But my experience tells me that even if we call the model twice, but with a smaller number of detections, it will still be slower than if we call it once for all detections, because it is necessary to transfer memory from the CPU to the GPU and back twice, and run the data through the model twice too. 

My final goal is to create tracker with ReID, which will work really fast, and the only idea i have is something like queue for embedding extraction, when we collect, for example, last 5 frames and all its detections corresponding to tracks (in case of EMA embeddings calculation), and we will calculate embeddings only at the moment, when we try to access them - let's call it lazy evaluation. And in case of two-stage trackers: on the first stage we calculate embeddings for comparison, on the second stage we add newly added detections to queue and extract embeddings only while we need to compare them and they are risky. Finally, this should not drop any metrices and give a small, but well-reasoned acceleration.

I am just sharing my experience in this task and would like to hear your opinions, maybe if you had the same or opposite experience.

**Comment by mikel-brostrom** [comment]
The motion based trackers have become so good for the high FPS case that ReID model are usually not bringing too much to the table. They are more important when FPS are very low and motion uncertainly is high

BoTSORT ablation
<img width="670" alt="Screenshot 2024-12-16 at 14 32 20" src="https://github.com/user-attachments/assets/79886613-17c6-4d51-b2da-e7fdb3a619c9" />

StrongSORT ablation
<img width="704" alt="Screenshot 2024-12-16 at 14 35 54" src="https://github.com/user-attachments/assets/b4c18bc1-4fdf-4749-a5f8-c9d0756eedae" />

**Comment by Fleyderer** [comment]
We add ReID -> FPS decreases -> ReID helps with lower FPS. But it still sounds like ReID solves the problem it created 🙂

At the same time I see that traditional motion based trackers are not using FPS as a parameter for math model and increasing/decreasing uncertainity, maybe it should be changed, I'll try to test it later.

It looks really tricky, because: If I have 25 FPS, this is good enough to use only motion information. If I have 8-12 FPS, I need ReID, but then I'll have 4-6 FPS and as a result it can work worse than without it, and existing metrics evaluation does not mean anything, because it is not about real-time.

For correct evaluation, for example, we firstly need to evaluate FPS of our tracker for each video, and then reduce the FPS of the streams, like if our original video FPS is 25 and tracker FPS is 20, drop every 5th frame and evaluate performance.

**Comment by mikel-brostrom** [comment]
ReID models using TensorRT backbone are FAST. I would like to have an experiment table with these results, but I don't 😢 . Motion uncertainty can also be there even if FPS are high.

> For correct evaluation, for example, we firstly need to evaluate FPS of our tracker for each video, and then reduce the FPS of the streams, like if our original video FPS is 25 and tracker FPS is 20, drop every 5th frame and evaluate performance.

This is a cool experiment

**Comment by Fleyderer** [comment]
> ReID models using TensorRT backend are FAST.

Yes, I will test final FPS using Triton Server, but sometimes i will need to inference on non-NVIDIA devices.

> Motion uncertainty can also be there even if FPS are high.

Do you mean situations where we track fast objects or objects with almost random trajectories? :thinking: I can't think of an example for person tracking with cases above, despite sports or another activities. But they are more specific, because sports requires uniform and ReID will not help, while uniform number detector would be the best solution.

> This is a cool experiment for validating FPS motion uncertainty adaptation. Sadly it will always be dataset dependent.

Why? Maybe I didn't write it very clearly, but I mean that we could evaluate on the MOT17/MOT20 dataset, but firstly evaluate FPS and then just drop some frames on the fly while evaluating HOTA/MOTA/IDF1, so we could know, what is better - use motion + ReID on 15 FPS or only motion on 25.

Additionally, with each day spent on this task, I see more and more incorrect solutions, measurements, processes. A simple example: we train the detector on one half of the video, and validate on the other. It's ridiculous. It is clear that the detector is not an end in itself, but the overall task consists of two parts, and it is also impossible to look at these HOTA/MOTA as the final metrics. But at the same time we can't just measure detector and tracker independently and say: "detector is great, tracker is great, let's just merge them".

**Comment by mikel-brostrom** [comment]
> Do you mean situations where we track fast objects or objects with almost random trajectories?

Or just objects that have large displacements in the image plane despite high FPS.

> Additionally, with each day spent on this task, I see more and more incorrect solutions, measurements, processes. A simple example: we train the detector on one half of the video, and validate on the other. It's ridiculous.

I agree with you here in that the processes are overly skewed and ultra-focused on MOT17/MOT20. That is why I "froze" the detections for the evaluations on this repo. I just do not care enough about the detectors I guess 🤣 .

**Comment by pialin** [comment]
hi,  has this Selective Mechanism been applied to all trackers?  how can i try it? Thank you @mikel-brostrom
